lib/cfa/matcher.rb

```diff
@@ -1,9 +1,32 @@
 module CFA
-  # Class used to create matcher, that allows to find specific option in augeas
-  # tree or subtree
-  # TODO: examples of usage
+  # The Matcher is used as a predicate on {AugeasElement}.
+  #
+  # Being a predicate, it is passed  to methods such as Enumerable#select
+  # or Array#index, returning a Boolean meaning whether a match was found.
+  #
+  # Acting on {AugeasElement} means it expects a Hash `e`
+  # containing `e[:key]` and `e[:value]`.
+  #
+  # It is used with the `&` syntax which makes the matcher
+  # act like a block/lambda/Proc (via {Matcher#to_proc}).
+  #
+  # @example
+  #    elements = [
+  #                {key: "#comment[]", value: "\"magical\" mostly works"},
+  #                {key: "DRIVE",      value: "magical"},
+  #                {key: "#comment[]", value: "'years' or 'centuries'"},
+  #                {key: "PRECISION",  value: "years"}
+  #               ]
+  #    drive_matcher = Matcher.new(key: "DRIVE")
+  #    i = elements.index(&drive_matcher)        # => 1
   class Matcher
-    # @block_yield matcher based on block. block gets two params, key and value
+    # The constructor arguments are constraints to match on an element.
+    # All constraints are optional.
+    # All supplied constraints must match, so it is a conjunction.
+    # @param key           [Object,nil] foo
+    # @param collection    [Object,nil] bar
+    # @param value_matcher [Object,nil] baz
```

Comment:
1. It used to be `===` but rubocop complains
2. yes, actually I need it for grub2, but found that using block form easier like `Matcher.new { |k, v| k.start_with? "ab" }`
for value it make especial sense for collection of comments like
`Matcher.new (collection: "#comment", value_matcher: /#{key}.*=.*/)` to get commented out value